### ISSUE TITLE
Combine positional and flag based arguments

### ISSUE THREAD

**Issue by xldenis** [Gabriel439/Haskell-Optparse-Generic-Library#18]
I'm trying to combine both positional and flag based arguments, I've tried embedding a record in a normal constructor like so:

```
data Config = Command [String] FlagArguments
data FlagArguments = Args {a :: Bool }
```

While the parser is correct, the help message is fairly useless:

```
Usage: proj-exe build [STRING] FlagArguments
```

What I'd like to get is:

```
Usage: proj-exe build [STRING] --a
```

I don't understand haskell generic programming well enough to provide the correct instances though.

**Comment by Gabriella439** [comment]
I think the fix here is to change the instances for tuples from this:

``` haskell
instance (ParseFields a, ParseFields b) => ParseRecord (a, b)
```

... to this:

``` haskell
instance (ParseRecord a, ParseRecord b) => ParseRecord (a, b)
```

... and likewise for other tuple instances.

Then the types `a` and `b` would be `Config` and `FlagArguments` in your case and it would do the right thing.

How does that sound?

**Comment by xldenis** [comment]
Hmm, I don't think that would really solve my issue, I just realized that my example might not have been the best.

A more accurate example could look something like this: 

```
data Config = Build [String] FlagArguments
            | Other [OtherArgs]
data FlagArguments = ...
```

though I suppose I could restructure it so that `Build` holds a tuple. :confused:

**Comment by Gabriella439** [comment]
Yeah, in general what you want to do is probably not going to work.  For example, consider the case where `FlagArguments` was a labeled field of `Build` or if it were a sum type

**Comment by Gabriella439** [comment]
For cases like that you will probably have to hand-write the top-level parser to do what you want

**Comment by xldenis** [comment]
Addressing both examples you've given:
1. If it is a labeled field then we can't generate a meaningful parser since in this case the important distinction is that `FlagArguments` is itself a record. 
2. Nested sum types pose a different, equally interesting problem: If `FlagArguments` is a sumtype, then I would expect `app build --help` to display style of usage message you would get for `app --help`

Currently, it seems like any nested type gets displayed as `TYPENAME` in help messages with no further explanation, making it totally useless.

Both nested subcommands and positional + flag based arguments are important for cli tools, I'd be happy to take a stab at implementing these if you feel it's worth supporting.

**Comment by Gabriella439** [comment]
If you think there is a clean way to implement this then I'll be happy to review it
